// res-pareto-foreach/dist/implementation.generated.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.$a = void 0;
const decorateDictionaryEntriesWithKey_p_1 = require("./implementations/decorateDictionaryEntriesWithKey.p");
const createDictionaryForEach_p_1 = require("./implementations/createDictionaryForEach.p");
const createArrayForEach_p_1 = require("./implementations/createArrayForEach.p");
const createEnrichedDictionaryForEach_p_1 = require("./implementations/createEnrichedDictionaryForEach.p");
const createEnrichedArrayForEach_p_1 = require("./implementations/createEnrichedArrayForEach.p");
__exportStar(require("./api"), exports);
exports.$a = {
    decorateDictionaryEntriesWithKey: decorateDictionaryEntriesWithKey_p_1.$$,
    createArrayForEach: createArrayForEach_p_1.$$,
    createDictionaryForEach: createDictionaryForEach_p_1.$$,
    createEnrichedArrayForEach: createEnrichedArrayForEach_p_1.$$,
    createEnrichedDictionaryForEach: createEnrichedDictionaryForEach_p_1.$$,
};
