// pareto-core-state/dist/functions/createStack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStack = void 0;
function createStack(seed) {
    const imp = [];
    seed.forEach(($) => {
        imp.push($);
    });
    //reversing, because the first element should be on top
    imp.reverse();
    return {
        pop: (onNotEmpty, onEmpty) => {
            const current = imp.pop();
            if (current === undefined) {
                return onEmpty();
            }
            else {
                return onNotEmpty(current);
            }
        },
        push: (v) => {
            imp.push(v);
        },
        isEmpty: () => {
            return imp.length === 0;
        },
        empty: (cb) => {
            while (true) {
                const entry = imp.pop();
                if (entry === undefined) {
                    return;
                }
                cb(entry);
            }
        },
        doUntil: ($c) => {
            while (true) {
                const res = imp.pop();
                if (res === undefined) {
                    return;
                }
                else {
                    imp.push(res);
                    const goOn = $c(res);
                    if (!goOn) {
                        //callback()
                        return;
                    }
                }
            }
        },
    };
}
exports.createStack = createStack;
